### PR TITLE
nrf52_bsim: Add support for UART & related fixes to drivers for simulation & related changes in tests

### PR DESCRIPTION
**Pull request by aescolar**
This PR includes 4 sets of related changes:
1. Take in the latest HW models which include the UART & UARTE peripherals, and do the associated changes in the nrf*bsim boards to allow using them (keep the DTS structure and add sensible defaults ; add a few headers for the board the UART driver needs
2. Minor changes to the UART and counter drivers to enable them running in simulation, and 2 ISR prototype fixes
3. Add 2 basic BLE tests using the UART, with the app + host in one nrf52 device, and the controller in another, communicating over UART. One test uses the hci_uart sample, the other hci_uart_async.
4. For the UART tests, add overlays where needed so they can also be run in the nrf52_bsim, and 2 minor fixes.

Also as minor related change, the native logger backend is now enabled even if SERIAL is used, as this seems much closer to what users expect (to see by default the logger output, also, in the invoking shell stdout even if the UART is enabled), and is something twister needs while running "native" target tests.

Note that the UART tests can be run and pass fine in the nrf52_bsim. You just need to loopback the Tx to the Rx:
```
cmake -GNinja -DBOARD=nrf52_bsim ../tests/drivers/uart/uart_async_api/  && ninja
zephyr/zephyr.exe -uart0_fifob_rxfile=cross.fifo -uart0_fifob_txfile=cross.fifo -v=2
```

The new UART peripheral documentation can be found in https://github.com/BabbleSim/ext_nRF_hw_models/blob/main/docs/UART.md
Note that for the nrf53, as we lack a GPIO peripheral model, which the UART driver needs, so we cannot yet use the UART.

### CODE REVIEW

File: boards/posix/nrf_bsim/soc/pinctrl_soc.h [hunk 1]
```diff
@@ -0,0 +1,13 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
```
Comment (Thalley) [inline boards/posix/nrf_bsim/soc/pinctrl_soc.h:2]: Should these be 2024 now?

File: drivers/serial/uart_nrfx_uarte.c [hunk 2]
```diff
@@ -1745,6 +1747,11 @@ static int uarte_instance_init(const struct device *dev,
 
 	data->dev = dev;
 
+#ifdef CONFIG_ARCH_POSIX
+	/* For simulation the DT provided peripheral address needs to be corrected */
+	((struct pinctrl_dev_config *)cfg->pcfg)->reg = (uintptr_t)cfg->uarte_regs;
```
Comment (Thalley) [inline drivers/serial/uart_nrfx_uarte.c:1752]: Is ARCH_POSIX always considered simulation?

File: tests/bsim/bluetooth/ll/conn/prj_split_uart.conf [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+CONFIG_BT_PERIPHERAL=y
+CONFIG_BT_PRIVACY=y
+CONFIG_BT_SMP=y
+CONFIG_BT_SIGNING=y
+CONFIG_BT_BAS=y
+CONFIG_BT_HRS=y
+CONFIG_BT_ATT_PREPARE_COUNT=2
+CONFIG_BT_GATT_CLIENT=y
+CONFIG_BT_L2CAP_DYNAMIC_CHANNEL=y
+CONFIG_BT_DEVICE_NAME="bsim_test_split"
```
Comment (Thalley) [inline tests/bsim/bluetooth/ll/conn/prj_split_uart.conf:13]:
```suggestion
CONFIG_BT_DEVICE_NAME="bsim_test_split_uart"
```

**Review by Thalley — DISMISSED** [review]

**Review by cvinayak — DISMISSED** [review]

**Review by alwa-nordic — DISMISSED** [review]

File: drivers/serial/uart_nrfx_uarte.c (hunk 2 above)
Comment (aescolar) [reply drivers/serial/uart_nrfx_uarte.c:1752]: "Yes" :). I mean this driver will be compiled for CONFIG_ARCH_POSIX only when built for the nrf*_bsim boards.
And in any case we are calling the new native_sim "simulator" too.. I guess it is a matter of convention.
(Wrt the need to replace registers addresses from hardcoded values to wherever they happen to be placed, that exists today for all posix arch targets)

**Review by alwa-nordic — APPROVED** [review]

**Review by Thalley — APPROVED** [review]

**Review by jori-nordic — APPROVED** [review]

File: drivers/serial/uart_nrfx_uarte.c [hunk 4]
```diff
@@ -1501,6 +1502,7 @@ static void uarte_nrfx_poll_out(const struct device *dev, unsigned char c)
 			}
 
 			irq_unlock(key);
+			Z_SPIN_DELAY(2);
```
Comment (nordic-krch) [inline drivers/serial/uart_nrfx_uarte.c:1505]: is this spin delay something specific to the posix?

Comment (aescolar) [reply drivers/serial/uart_nrfx_uarte.c:1505]: It defines out to nothing for other architectures.

**Review by nordic-krch — APPROVED** [review]

**Pull request merged**